Content.Server/Weapons/Ranged/Systems/GunSystem.cs

```diff
@@ -132,6 +133,28 @@ private void OnBallisticPrice(EntityUid uid, BallisticAmmoProviderComponent comp
                 case CartridgeAmmoComponent cartridge:
                     if (!cartridge.Spent)
                     {
+                        if (gun.CompatibleAmmo != null &&
+                            !gun.CompatibleAmmo.Exists(ammoAllowed => ammoAllowed.Equals(cartridge.Prototype))
+                            && user != null)
+                        {
+                            var damage = new DamageSpecifier();
+                            if (gun.DamageOnWrongAmmo != null)
+                                Damageable.TryChangeDamage(user, gun.DamageOnWrongAmmo, origin: user);
+                            _stun.TryParalyze(user.Value, TimeSpan.FromSeconds(3f), true);
+
+                            Audio.PlayPvs(new SoundPathSpecifier("/Audio/Weapons/Guns/Gunshots/bang.ogg"), gunUid);
+
+                            PopupSystem.PopupEntity(Loc.GetString("Wrong ammo!"), user.Value);
```

Comment:
loc.getstring should be passed a locale id which is stored in Resources/Locale/...

if the ftl looks like `gun-component-wrong-ammo = Wrong ammo!` then use `Loc.GetString("gun-component-wrong-ammo")`